sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/Hotspots/UsingCommandLineArgumentsBase.cs

```diff
@@ -0,0 +1,41 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UsingCommandLineArgumentsBase<TSyntaxKind> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+    {
+        protected const string DiagnosticId = "S4823";
+        protected const string MessageFormat = "Make sure that command line arguments are used safely here.";
+
+        protected MethodDeclarationTracker<TSyntaxKind> MethodDeclarationTracker { get; set; }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            MethodDeclarationTracker.Track(context,
+                MethodDeclarationTracker.MatchSimpleNames("Main"),
+                MethodDeclarationTracker.IsMainMethod(),
+                MethodDeclarationTracker.ParameterAtIndexIsUsed(0));
+        }
```

Comment:
I like this way of defining rules: short, declarative, easy to understand. No doubt we could still make improvements, but I thinks it is better than the approach of having each rule directly registering callbacks against the supplied analyzer context.
Let's hope it doesn't have a negative impact on performance...

Reply:
Perhaps these are the slowest rules we have ;)